src/main/java/net/sf/jabref/gui/entryeditor/FieldExtraComponents.java

```diff
@@ -0,0 +1,258 @@
+        button.setToolTipText(ABBREVIATION_TOOLTIP_TEXT);
+        button.addActionListener(new ActionListener() {
+
+            @Override
+            public void actionPerformed(ActionEvent actionEvent) {
+                String text = editor.getText();
+                if (Abbreviations.journalAbbrev.isKnownName(text)) {
+                    String s = Abbreviations.toggleAbbreviation(text);
+
+                    if (s != null) {
+                        editor.setText(s);
+                        storeFieldAction.actionPerformed(new ActionEvent(editor, 0, ""));
+                        panel.undoManager.addEdit(new UndoableFieldChange(entry, editor.getFieldName(), text, s));
+                    }
+                }
+            }
+        });
+
+        controls.add(button, BorderLayout.SOUTH);
+        return controls;
+    }
+
+    public static JComponent getBrowseExtraComponent(JabRefFrame frame, FieldEditor fieldEditor,
+            EntryEditor entryEditor) {
+        JButton but = new JButton(Localization.lang("Browse"));
+        ((JComponent) fieldEditor).addMouseListener(entryEditor.new ExternalViewerListener());
+
+        but.addActionListener(new ActionListener() {
+
+            @Override
+            public void actionPerformed(ActionEvent e) {
+                String dir = fieldEditor.getText();
+
+                if (dir.isEmpty()) {
+                    dir = Globals.prefs.get(fieldEditor.getFieldName() + Globals.FILETYPE_PREFS_EXT, "");
+                }
+
+                String chosenFile = FileDialogs.getNewFile(frame, new File(dir), '.' + fieldEditor.getFieldName(),
+                        JFileChooser.OPEN_DIALOG, false);
+
+                if (chosenFile != null) {
+                    File newFile = new File(chosenFile); // chooser.getSelectedFile();
+                    fieldEditor.setText(newFile.getPath());
+                    Globals.prefs.put(fieldEditor.getFieldName() + Globals.FILETYPE_PREFS_EXT, newFile.getPath());
+                    entryEditor.updateField(fieldEditor);
+                }
+            }
+        });
+
+        return but;
+    }
+
+    static JComponent getBrowseDocExtraComponent(JabRefFrame frame, BasePanel panel, FieldEditor fieldEditor,
+            EntryEditor entryEditor, Boolean isZip) {
+
+        final String ext = '.' + fieldEditor.getFieldName().toLowerCase();
+        final OpenFileFilter off;
+        if (isZip) {
+            off = new OpenFileFilter(new String[] {ext, ext + ".gz", ext + ".bz2"});
+        } else {
+            off = new OpenFileFilter(new String[] {ext});
+        }
+
+        return new ExternalFilePanel(frame, panel.metaData(), entryEditor, fieldEditor.getFieldName(), off,
+                fieldEditor);
+    }
+
+    public static JComponent getExternalExtraComponent(FieldEditor fieldEditor, EntryEditor entryEditor) {
+        ((JComponent) fieldEditor).addMouseListener(entryEditor.new ExternalViewerListener());
+
+        return null;
```

Comment:
Hm, I do not like that null is returned in some cases. Could this just be converted to returning an Optional? Would also apply to the getExtra method.